### PR TITLE
Modeler update + layout alg doc

### LINKED ISSUE #3458: [FEATURE][processing] Allow expression variables to be set for a model
Original commit: https://github.com/qgis/QGIS/commit/c1aac3228c44f5add412be77a5221af34fd1a671 by nyalldawson

This adds a new "Model Variables" dock panel to the model editor, allowing
users to create and set custom expression variables for use in the model.
These variables are available anywhere expressions are (correctly) evaluated
within the model, so can be used as input parameter values for child
algorithms, within data-defined dynamic parameters, etc.

The use case here is for models which use a constant value throughout
multiple steps within the model (e.g. @target_resolution: a target
raster resolution, @max_simplification: a simplification value for
input features coming from different sources, etc), allowing users
one single place to define and edit these constant values (instead
of hunting down and setting them in multiple places throughout the model).

These variables are stored within the model itself, and are not exposed
outside of the model designer dialog.

### LINKED ISSUE #4964: Request in QGIS (Comments in Processing Models)
### Request for documentation
From pull request QGIS/qgis#34856
Author: @nyalldawson
QGIS version: 3.14

**Comments in Processing Models**

### PR Description:
(includes https://github.com/qgis/QGIS/pull/34852)

This allows users to create comments attached to model components (inputs, algorithms or outputs). Comments are shown linked to the associated component, and can be freely moved around the model.

(will be more user-friendly when model components can also be resized :laughing: )

Peek 2020-03-04 16-27

 Funded by Fisel + König

### Commits tagged with [need-docs] or [FEATURE]
"[FEATURE][processsing] Add support for comments attached to components\n\nThis allows users to create comments attached to model components (inputs,\nalgorithms or outputs). Comments are shown linked to the associated component,\nand can be freely moved around the model."

### LINKED ISSUE #4966: Request in QGIS (Remember parameter values between model designer runs)
### Request for documentation
From pull request QGIS/qgis#34895
Author: @nyalldawson
QGIS version: 3.14

**Remember parameter values between model designer runs**

### PR Description:
When designing a model, users typically will need to run the model many times as they tweak its structure.

This change causes the parameters used when running the model from the designer to be remembered and saved into the model, so that each time you run the model from the designer you don't have to re-set all the input parameter values to the desired test ones.

Makes iterative model design SO much easier!

Sponsored by Alta Ehf

Also includes a partial port of the model designer dialog from Python to c++, because c++ >> Python.

### Commits tagged with [need-docs] or [FEATURE]
"[FEATURE][processing] Remember parameter values between model designer runs\n\nWhen designing a model, users typically will need to run the model\nmany times as they tweak its structure.\n\nThis change causes the parameters used when running the model from\nthe designer to be remembered and saved into the model, so that\neach time you run the model from the designer you don't have to\nre-set all the input parameter values to the desired test ones.\n\nMakes iterative model design SO much easier!\n\nSponsored by Alta Ehf"

### LINKED ISSUE #5003: [FEATURE][processing] Add action to model designer to delete all selected components
Original commit: https://github.com/qgis/QGIS/commit/1e67ee42fa95d7ea3078f346f28a5d29b77c2177 by nyalldawson

Unfortunately this naughty coder did not write a description... :-(

### LINKED ISSUE #5030: Request in QGIS (Add "filter by geometry type" and "filter by layer type" algorithms to processing)
### Request for documentation
From pull request QGIS/qgis#35130
Author: @nyalldawson
QGIS version: 3.14

**Add "filter by geometry type" and "filter by layer type" algorithms to processing**

### PR Description:
This PR adds two new algorithms to processing:

- "Filter by geometry type": This algorithm filters features by their geometry type. Incoming features will be directed to different outputs based on whether they have a point, line or polygon geometry. It allows for model creation which responds to different input layer geometry types by applying different logic depending on the input geometry type.

- "Filter by layer type": This algorithm allows conditional model branching based on an input
layer type. For instance, it allows a model to adapt to the actual layer type of a generic "map layer" parameter input, and decide which branch of the model to run as a result.

It also adds in the required api to allow algorithms to "prune" model branches based on their calculated results. E.g. a model which returns the new FlagPruneModelBranchesBasedOnAlgorithmResults flag will cause any remaining parts of the model which are dependent on the outputs of that algorithm to be entirely skipped IF the algorithm does not return that particular output. (This is a prerequisite component for a future generic "conditional branching by expression" algorithm, and also used by "filter by layer type" to control the model flow based on the input layer type)

Refs NRCan Contract#3000707093

### Commits tagged with [need-docs] or [FEATURE]
"[FEATURE][processing] Add new algorithm \"Filter by geometry type\"\n\nThis algorithm filters features by their geometry type. Incoming\nfeatures will be directed to different outputs based on whether\nthey have a point, line or polygon geometry.\n\nAllows for model creation which responds to different input\nlayer geometry types by applying different logic depending on the\ninput geometry type." "[FEATURE][processing] Add algorithm to filter by layer type\n\nThis algorithm allows conditional model branching based on an input\nlayer type. For instance, it allows a model to adapt to the actual\nlayer type of a generic \"map layer\" parameter input, and decide\nwhich branch of the model to run as a result."

### LINKED ISSUE #5189: Add "Save Log to File" algorithm for models  (Request in QGIS)
### Request for documentation
From pull request QGIS/qgis#35473
Author: @nyalldawson
QGIS version: 3.14

**Add "Save Log to File" algorithm for models **

### PR Description:
This algorithm saves the contents of the execution log (right up to the point in the model at which the 'save log' algorithm executes) to a file.

It can be used to automatically store the debugging log when running models for later reference and transparency.

Refs NRCan Contract#3000707093

### Commits tagged with [need-docs] or [FEATURE]
"[processing][FEATURE] Add \"Save Log to File\" algorithm for models\n\nThis algorithm saves the contents of the execution log (right up to\nthe point in the model at which the 'save log' algorithm executes)\nto a file.\n\nIt can be used to automatically store the debugging log when running\nmodels for later reference and transparency."

### LINKED ISSUE #5191: Processing: show input and output values for children after running model through designer (Request in QGIS)
### Request for documentation
From pull request QGIS/qgis#35474
Author: @nyalldawson
QGIS version: 3.14

**Processing: show input and output values for children after running model through designer**

### PR Description:
A picture explains this best! After running the algorithm through the designer, you now see the values calculated for the inputs and outputs for each child algorithm:

image

This is very useful for debugging models -- you can see a much clearer visual picture of the flow of values through the model.

Refs NRCan Contract#3000707093

### Commits tagged with [need-docs] or [FEATURE]
"[processing][FEATURE] After running a model through the model designer,\nshow the obtained values for all child algorithm outputs within the designer\ncanvas\n\nThis gives users better tools for debugging models - they can see exactly\nwhat values were output by the child algorithms and flowed into other\nparts of their model"

### LINKED ISSUE #5281: [FEATURE] Add action to manually validate a processing model
Original commit: https://github.com/qgis/QGIS/commit/fd9815809cf6636db7b7c8245def57f2677944b5 by nyalldawson

Allows users to run a validation over their model, listing any issues
found with the model

### LINKED ISSUE #5294: [FEATURE][processing] New modeler algorithm for conditional branches in models
Original commit: https://github.com/qgis/QGIS/commit/6a1b10686b84b681225e6acc4d498f9d18d7627c by nyalldawson

This algorithm allows useres to setup multiple conditions (via qgis expressions),
which cause their corresponding branch of the model to be run or skipped
depending on the result of the expression evaluation.

Sponsored by Andreas Neumann

### LINKED ISSUE #5780: [FEATURE][processing] Add algorithms to export a print layout as PDF or image
Original commit: https://github.com/qgis/QGIS/commit/f75a9d21d5aa98221d2d32a7db1321000ba56c2c by nyalldawson

This allows for models which export print layouts from the current project. One
use case for this is allowing users to create an in-project model which exports
a particular set of layouts from the project to certain folders, so that they
can easily re-export the current project in a single operation instead of
having to manually open multiple layouts and export one-by-one.

Additionally, with the new capabilities to have expression based output files
inside models, you can automatically export the layouts to a folder with
the current date tag and include this in the exported file names!

### LINKED ISSUE #5781: [feature][processing] Add modeler algorithm to create directories
Original commit: https://github.com/qgis/QGIS/commit/da8fa08af5283f9b338bbe7b2ae63e67090570ee by nyalldawson

This algorithm allows models to create new file directories. Useful for
models which have to output results to certain dynamically created paths
(e.g. a folder containing the current date)

### LINKED ISSUE #5968: [FEATURE][processing] Export atlas layout as image algorithm (Request in QGIS)
### Request for documentation
From pull request QGIS/qgis#38067
Author: @nirvn
QGIS version: 3.16

**[FEATURE][processing] Export atlas layout as image algorithm**

## Description

This PR adds a new native processing algorithm to export layout atlases. Obligatory screenshot:
Screenshot from 2020-07-31 16-07-27

This is _super_ useful in models where remote datasets are fetched, processed, with the goal of being used as atlas coverage layers.

### Commits tagged with [need-docs] or [FEATURE]
"[FEATURE][processing] Export print layout atlas as image(s) algorithm"

### LINKED ISSUE #6007: [FEATURE][processing] Export layout atlas as PDF algorithm (Request in QGIS)
### Request for documentation
From pull request QGIS/qgis#38196
Author: @nirvn
QGIS version: 3.16

**[FEATURE][processing] Export layout atlas as PDF algorithm**

## Description

PDF counterpart to the recent export atlas layout to image algorithm.

### Commits tagged with [need-docs] or [FEATURE]
"[FEATURE][processing] Export layout atlas as PDF algorithm"

### PR DESCRIPTION
**Pull request by roya0045**
Refreshing of the modeler section and dedicated algorythms.

Fixes #5781 Fixes #5294 Fixes #5189 Fixes #5030

Fixes #4966 Fixes #4964 Fixes #3458 Fixes #5191 Fixes #5281

Fixes #5780

Extra: Fixes #6007 Fixes #5968

### CODE REVIEW

**Comment by DelazJ** [comment]
@roya0045 I know it's wip but few remarks:
- According to [guidelines](https://docs.qgis.org/testing/en/docs/documentation_guidelines/writing.html), description doesn't begin with "this algoritm..."
- you'll find in the guides some other rules about anchor construction, list of "substitution" to use for input/output data type
- algorithms are alphabetically sorted in the page
- do not forget default/possible values if any/list
- the feature filter algorithm is already documented in vector table

Hope that helps to progress...

**Comment by roya0045** [comment]
@DelazJ This PR has enough related issue. Review at your conveinience, thanks!

File: docs/user_manual/processing_algs/qgis/cartography.rst [hunk 1]
```diff
@@ -421,7 +421,478 @@ Outputs
...
+       Optional
+     - ``LAYERS``
+     - [enumeration] [layer]
+     - Layers to display in the map item(s) whose contents are not locked
+   * - **Image format**
+     - ``EXTENSION``
+     - [list]
+
+       Default: png
+     - File format of the generated output(s).
```
Comment (DelazJ) [inline docs/user_manual/processing_algs/qgis/cartography.rst:500]:
```suggestion
     - File format of the generated output(s).
       The list of available formats varies depending on OS and installed drivers.
```
I don't have same list on macOS and Windows, hence the assumption above.

File: docs/user_manual/processing_algs/qgis/cartography.rst [hunk 2]
```diff
@@ -421,7 +421,478 @@ Outputs
...
+
+.. list-table::
+   :header-rows: 1
+   :widths: 20 20 20 40
+
+   * - Label
+     - Name
+     - Type
+     - Description
+   * - **Image file**
```
Comment (DelazJ) [inline docs/user_manual/processing_algs/qgis/cartography.rst:887]:
```suggestion
   * - **PDF file**
```
I missed this one

**Review by DelazJ — COMMENTED** [review]
I pushed a couple of commits fixing build and issues relative to guidelines observance and completing algs parameters. Here some comments I'm not 100% sure of.
PS: I did not review the modeler tools as I'm not really comfortable with/used to this area, and how their parameters are exposed.

File: docs/user_manual/processing/modeler.rst [hunk 3]
```diff
@@ -265,6 +263,22 @@ In order to use the algorithm from the toolbox, it has to be saved
 and the modeler dialog closed, to allow the toolbox to refresh its
 contents.
 
+When using the editor to execute a model, any non-default value will be 
+saved in the inputs. This mean that executing the model at a later time from
+the editor will have the dialog prefilled with those value on any subsequent run.
```
[the remark this answers is not in the document]
Comment (roya0045) [reply docs/user_manual/processing/modeler.rst:268]: All the non defult values will be reused. Iwas not sure if it was better to stay singular and go on an input basis or a model/dialog basis.

File: docs/user_manual/processing_algs/qgis/cartography.rst (hunk 1 above)
Comment (roya0045) [reply docs/user_manual/processing_algs/qgis/cartography.rst:500]: This is possible, I am not sure if things are so distinct, I assume all platforms have access to the most common formats at least.

**Comment by roya0045** [comment]
@DelazJ  Thanks for the cleanup, is there a way to get the tags and other elements sorted out? For the doc I mostly work in a code editor and there might be a better way to do things with antocompletion or must every tag and such be done manually? Thanks,

File: docs/user_manual/processing/modeler.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (DelazJ) [reply docs/user_manual/processing/modeler.rst]: Took the dialogue way...

**Comment by DelazJ** [comment]
> is there a way to get the tags and other elements sorted out? For the doc I mostly work in a code editor and there might be a better way to do things with autocompletion or must every tag and such be done manually?

No automatic tool I'm aware of. I do it manually with a text editor that autocompletes based on what is already in the file; no relation with code base.

**Comment by roya0045** [comment]
@DelazJ Thanks for the reply and getting this over the line!

**Comment by roya0045** [comment]
@DelazJ Symbology (as a whole or subdivided by data source) seems to be the next best block. I'll see if I can start work on that and hopefully cause you less work.

**Pull request merged**